internal/sql/pg/system.go

```diff
@@ -198,6 +202,452 @@ var settingValidations = map[string]settingValidFn{
 	"server_encoding": wantStringFn("UTF8"),
 }
 
+func queryRowFunc(ctx context.Context, conn *pgx.Conn, sql string,
+	scans []any, fn func() error, args ...any) error {
+	rows, _ := conn.Query(ctx, sql, args...)
+	_, err := pgx.ForEachRow(rows, scans, fn)
+	return err
+}
+
+// QueryRowFunc will attempt to execute an SQL statement, handling the rows and
+// returned values as described by the sql.QueryScanner interface. If the
+// provided Executor is also a sql.QueryScanner, that method will be used,
+// otherwise, it will attempt to use the underlying DB connection. The latter is
+// supported for all concrete transaction types in this package as well as
+// instances of the pgx.Tx interface.
+func QueryRowFunc(ctx context.Context, tx sql.Executor, stmt string,
+	scans []any, fn func() error, args ...any) error {
+	switch ti := tx.(type) {
+	case sql.QueryScanner:
+		return ti.QueryScanFn(ctx, stmt, scans, fn, args...)
+	case conner:
+		conn := ti.Conn()
+		return queryRowFunc(ctx, conn, stmt, scans, fn, args...)
+	}
+	return errors.New("cannot query with scan values")
+}
+
+// FieldDesc describes result value from a query. This is used to convey
+// information on the values passed to the closure given to QueryRowFuncAny.
+type FieldDesc struct {
```

Comment:
Unexport? This gives the caller access to the OID, which for the uint256 domain, will be different on every system. Would hate for someone to forget this and rely on this functionality externally later

Reply:
Ah, yes!  So, the motivation is poorly explained, but this pertains to the main similarity between `QueryRowFuncAny` and the current `Execute`/`query` functions, which also use `row.Values()`.  Essentially `QueryRowFuncAny` should behave like `query` but does something for each row rather than loading all rows into memory.

This `FieldDesc` relates to the above since it is an argument to the function that is executed for each row.

So yeah, both `FieldDesc` and `QueryRowFuncAny` in their current forms should be unexported.  Any exported `QueryRowFuncAny` should not trouble the user with OIDs and such, and should ensure the `[]any` contains the same types of values that the exported `Execute` methods return. (needs to use the registered `datatype`s + `decodeFromPG` => the `Decode` methods)